ballerina/channel_csv_io.bal

```diff
@@ -52,43 +52,74 @@ isolated function channelWriteCsv(string path, FileWriteOption option, string[][
+isolated function channelWriteCsvFromStream(string path, FileWriteOption option, stream<string[]|map<anydata>, Error?> csvStreamToWrite) returns
 Error? {
-    WritableCSVChannel csvChannel = check getWritableCSVChannel(writableChannel);
-    do {
-        if csvStream is stream<string[], Error?> {
-            record {|string[] value;|}|Error? csvRecordString = csvStream.next();
+    if csvStreamToWrite is stream<string[], Error?> {
+        WritableCSVChannel csvChannel = check getWritableCSVChannel(check openWritableCsvFile(path, option = option));
+        do {
+            record {|string[] value;|}|Error? csvRecordString = csvStreamToWrite.next();
             while csvRecordString is record {|string[] value;|} {
                 check csvChannel.write(csvRecordString.value);
-                csvRecordString = csvStream.next();
+                csvRecordString = csvStreamToWrite.next();
+            }
+        } on fail Error err {
+            check csvStreamToWrite.close();
+            check csvChannel.close();
+            return err;
+        }
+        check csvChannel.close();
+    } else if csvStreamToWrite is stream<map<anydata>, Error?> {
+        boolean skipHeaders = false;
+        string[] headersFromStream;
+        record {|map<anydata> value;|}? csvRecordMap = check csvStreamToWrite.next();
+        do {
+            if csvRecordMap !is () {
+                headersFromStream = csvRecordMap["value"].keys();
+            } else {
+                check csvStreamToWrite.close();
+                return;
+            }
+        } on fail Error err {
+            check csvStreamToWrite.close();
+            return err;
+        }
+        if option == APPEND {
+            string[] headersFromCSV = check readHeadersFromCsvFile(path);
+            headersFromCSV = check validateCsvHeaders(headersFromCSV, headersFromStream);
+            if headersFromCSV.length() != 0 {
+                skipHeaders = true;
+                headersFromStream = headersFromCSV;
+            }
+        }
+        WritableCSVChannel csvChannel = check getWritableCSVChannel(check openWritableCsvFile(path, option = option));
+        do {
+            if !skipHeaders {
+                check csvChannel.write(headersFromStream);
             }
-        } else if csvStream is stream<map<anydata>, Error?> {
-            record {|map<anydata> value;|}? csvRecordMap = check csvStream.next();
             while csvRecordMap is record {|map<anydata> value;|} {
                 string[] sValues = [];
-                foreach [string, anydata] [_, value] in csvRecordMap["value"].entries() {
-                    sValues.push(value.toString());
+                foreach string header in headersFromStream {
+                    sValues.push(csvRecordMap["value"].get(header).toString());
                 }
+                csvRecordMap = check csvStreamToWrite.next();
                 check csvChannel.write(sValues);
```

Comment:
Is this the correct order? We are iterating the stream even before the first element write.